internal/consensus/state_data.go

```diff
@@ -0,0 +1,467 @@
+	if s.CommitRound > -1 && 0 < s.Height && s.Height != state.LastBlockHeight {
+		panic(fmt.Sprintf(
+			"updateToState() expected state height of %v but found %v",
+			s.Height, state.LastBlockHeight,
+		))
+	}
+
+	if !s.state.IsEmpty() {
+		if s.state.LastBlockHeight > 0 && s.state.LastBlockHeight+1 != s.Height {
+			// This might happen when someone else is mutating cs.state.
+			// Someone forgot to pass in state.Copy() somewhere?!
+			panic(fmt.Sprintf(
+				"inconsistent cs.state.LastBlockHeight+1 %v vs cs.Height %v",
+				s.state.LastBlockHeight+1, s.Height,
+			))
+		}
+		if s.state.LastBlockHeight > 0 && s.Height == s.state.InitialHeight {
+			panic(fmt.Sprintf(
+				"inconsistent cs.state.LastBlockHeight %v, expected 0 for initial height %v",
+				s.state.LastBlockHeight, s.state.InitialHeight,
+			))
+		}
+
+		// If state isn't further out than cs.state, just ignore.
+		// This happens when SwitchToConsensus() is called in the reactor.
+		// We don't want to reset e.g. the Votes, but we still want to
+		// signal the new round step, because other services (eg. txNotifier)
+		// depend on having an up-to-date peer state!
+		if state.LastBlockHeight < s.state.LastBlockHeight {
+			//s.logger.Debug(
+			//	"ignoring updateToState()",
+			//	"new_height", state.LastBlockHeight+1,
+			//	"old_height", s.state.LastBlockHeight+1)
+			//s.newStep()
+			return
+		}
+	}
+
+	// Reset fields based on state.
+	validators := state.Validators
+
+	switch {
+	case state.LastBlockHeight == 0: // Very first commit should be empty.
+		s.LastCommit = (*types.Commit)(nil)
+		s.LastPrecommits = (*types.VoteSet)(nil)
+	case s.CommitRound > -1 && s.Votes != nil && commit == nil: // Otherwise, use cs.Votes
+		if !s.Votes.Precommits(s.CommitRound).HasTwoThirdsMajority() {
+			panic(fmt.Sprintf(
+				"wanted to form a commit, but precommits (H/R: %d/%d) didn't have 2/3+: %v",
+				state.LastBlockHeight, s.CommitRound, s.Votes.Precommits(s.CommitRound),
+			))
+		}
+		s.LastPrecommits = s.Votes.Precommits(s.CommitRound)
+		s.LastCommit = s.LastPrecommits.MakeCommit()
+	case commit != nil:
+		// We either got the commit from a remote node
+		// In which Last precommits will be nil
+		// Or we got the commit from finalize commit
+		// In which Last precommits will not be nil
+		s.LastPrecommits = s.Votes.Precommits(s.CommitRound)
+		s.LastCommit = commit
+	case s.LastCommit == nil:
+		// NOTE: when Tendermint starts, it has no votes. reconstructLastCommit
+		// must be called to reconstruct LastPrecommits from SeenCommit.
+		panic(fmt.Sprintf(
+			"last commit cannot be empty after initial block (H:%d)",
+			state.LastBlockHeight+1,
+		))
+	}
+
+	// Next desired block height
+	height := state.LastBlockHeight + 1
+	if height == 1 {
+		height = state.InitialHeight
+	}
+
+	s.logger.Debug("updating state height", "newHeight", height)
+
+	// RoundState fields
+	s.updateHeight(height)
+	s.updateRoundStep(0, cstypes.RoundStepNewHeight)
```

Comment:
I moved `State.updateToState` into `StateData.updateToState` because I based on the idea - who is owner of the data, and since, we do not use external deps to update the state, I attached it to `StateData`.
The most important is - when and where we call `StateData.updateToState`, we do it in three places:
* Reactor.SwitchToConsensus
* State.updateStateFromStore
* ApplyCommitCommand

also, I ain't happy with this long function :-)